File: KRSU/116.py
import copy


def FIFO(arrival, release):
    arrival = copy.deepcopy(arrival)
    arrival[0][0] -= release
    while arrival and arrival[0][0] <= 0:
        previous, _ = arrival.pop(0)
        if arrival:
            arrival[0][0] += previous
    remain = 0
    while arrival and arrival[0][0]>0:
        weight, price = arrival.pop(0)
        remain += weight*price
    return remain


def LIFO(arrival, release):
    arrival = copy.deepcopy(arrival)
    arrival[-1][0] -= release
    while arrival and arrival[-1][0] <= 0:
        previous, _ = arrival.pop()
        if arrival:
            arrival[-1][0] += previous
    remain = 0
    while arrival and arrival[-1][0] > 0:
        weight, price = arrival.pop()
        remain += weight*price
    return remain


def MEDIUM(arrival, release):
    total_weight = sum(weight for weight, _ in arrival)
    medium = sum(weight*price for weight, price in arrival)//total_weight
    remaining_weight = total_weight - release
    return remaining_weight*medium

res = [[0, 'LIFO'], [0, 'FIFO'], [0, 'MEDIUM']]

for i in range(int(input())):
    K, M = map(int,input().split())
    arrival, release = [], []
    for k in range(K):
        w, p = map(int, input().split())
        arrival.append([w, p])
    for m in range(M):
        release.append(int(input()))
    release = int(sum(release))
    if release > 0:
        lifo, fifo, medium = LIFO(arrival[:], release), FIFO(arrival[:], release), MEDIUM(arrival[:], release)
        print(fifo, lifo, medium, sep='\n')
        res[0][0]+=lifo
        res[1][0]+=fifo
        res[2][0]+=medium
    else:
        print('0\n0\n0')
res.sort(key=lambda x: (x[0],x[1]))
print(res[0][1])

File: KRSU/116.cpp
#include <bits/stdc++.h>
using namespace std;
int LIFO(deque < pair <int, int> > arrival, int release){
    arrival.back().first-=release;
    int weight,price, remain=0;
    while (!arrival.empty() && arrival.back().first <=0){
        tie(weight,price) = (arrival.back());
        arrival.pop_back();
        if (!arrival.empty())
            arrival.back().first+=weight;
    }
    while (!arrival.empty() && arrival.back().first>0){
        tie(weight,price) = (arrival.back());
        arrival.pop_back();
        remain +=weight*price;
    }
    return remain;
}
int FIFO(deque < pair <int, int> > arrival, int release){
    arrival.front().first-=release;
    int weight,price, remain=0;
    while (!arrival.empty() && arrival.front().first <=0){
        tie(weight,price) = (arrival.front());
        arrival.pop_front();
        if (!arrival.empty())
            arrival.front().first+=weight;
    }
    while (!arrival.empty() && arrival.front().first>0){
        tie(weight,price) = (arrival.front());
        arrival.pop_front();
        remain +=weight*price;
    }
    return remain;
}
int MEDIUM(deque <pair<int,int>> arrival, int release){
    int total_weight=0, remaining_weight=0, total=0, medium;
    for (auto it : arrival) {
        total_weight += it.first;
        total+=it.first*it.second;
    }
    medium = total/total_weight;
    remaining_weight = total_weight-release;
    return medium*remaining_weight;

}
int main(){
int N,M,K,weight,price, temp, fifo, lifo, medium;
vector <pair <int, string>> ans = {
        {0,"FIFO"}, {0, "LIFO"}, {0,"MEDIUM"}
};
cin >> N;
for (int i =0;i<N;i++){
    cin >> M >> K;
    deque <pair<int,int>> arrival;
    int release=0;
    for (int m=0;m<M;m++){
        cin >> weight >> price;
        arrival.emplace_back(make_pair(weight,price));
    }
    for (int k=0;k<K;k++){
        cin >> temp;
        release+=temp;
    }
    fifo = FIFO(arrival, release); lifo = LIFO(arrival, release); medium = MEDIUM(arrival, release);
    cout << fifo << "\n" << lifo << "\n" << medium << "\n";
    ans[0].first+=fifo; ans[1].first += lifo; ans[2].first +=medium;

}
sort(ans.begin(), ans.end(),[] (const pair<int, string> &a,const pair<int, string> &b) -> bool{
    return a.first < b.first ? true : a.second < b.second;
});
cout << ans[0].second;

}
